src/drivers/dshot/DShot.cpp

```diff
@@ -239,18 +249,73 @@ void DShot::handle_new_telemetry_data(const int telemetry_index, const DShotTele
 		esc_status.esc_connectiontype = esc_status_s::ESC_CONNECTION_TYPE_DSHOT;
 		esc_status.esc_count = _telemetry->handler.numMotors();
 		++esc_status.counter;
-		// FIXME: mark all ESC's as online, otherwise commander complains even for a single dropout
-		esc_status.esc_online_flags = (1 << esc_status.esc_count) - 1;
-		esc_status.esc_armed_flags = (1 << esc_status.esc_count) - 1;
 
-		_telemetry->esc_status_pub.update();
-
-		// reset esc data (in case a motor times out, so we won't send stale data)
-		memset(&esc_status.esc, 0, sizeof(_telemetry->esc_status_pub.get().esc));
-		esc_status.esc_online_flags = 0;
+		ret = 1; // Indicate we wrapped, so we publish data
 	}
 
 	_telemetry->last_telemetry_index = telemetry_index;
+
+	return ret;
+}
+
+void DShot::publish_esc_status(void)
+{
+	esc_status_s &esc_status = _telemetry->esc_status_pub.get();
+
+	// clear data of the esc that are offline
+	for (int index = 0; (index < _telemetry->last_telemetry_index); index++) {
+		if ((esc_status.esc_online_flags & (1 << index)) == 0) {
+			memset(&esc_status.esc[index], 0, sizeof(struct esc_report_s));
+		}
+	}
+
+	// FIXME: mark all UART Telemetry ESC's as online, otherwise commander complains even for a single dropout
+	esc_status.esc_online_flags = (1 << esc_status.esc_count) - 1;
+	esc_status.esc_armed_flags = (1 << esc_status.esc_count) - 1;
+
+	if (_bidirectional_dshot_enabled) {
+		esc_status.esc_online_flags |= _output_mask;
+		esc_status.esc_armed_flags |= _output_mask;
+		esc_status.esc_count = _telemetry->handler.numMotors();
+
+		for (int index = 0; (index < esc_status.esc_count); index++) {
+			if (up_bdshot_channel_status(index) == 0) {
+				esc_status.esc_online_flags &= ~(1 << index);
+			}
+		}
+	}
+
+	// ESC telem wrap around or bdshot update
+	_telemetry->esc_status_pub.update();
+
+	// reset esc online flags
+	esc_status.esc_online_flags = 0;
+}
+
+int DShot::handle_new_bdshot_erpm(void)
+{
+	int num_erpms = 0;
+	int erpm;
+	uint8_t channel;
+	esc_status_s &esc_status = _telemetry->esc_status_pub.get();
+
+	esc_status.timestamp = hrt_absolute_time();
+	esc_status.counter = _esc_status_counter++;
+	esc_status.esc_connectiontype = esc_status_s::ESC_CONNECTION_TYPE_DSHOT;
+	esc_status.esc_armed_flags = _outputs_on;
+
+	for (channel = 0; channel < 8; channel++) {
+		if (up_bdshot_get_erpm(channel, &erpm) == 0) {
+			num_erpms++;
+			esc_status.esc_online_flags |= 1 << channel;
+			esc_status.esc[channel].timestamp = hrt_absolute_time();
+			esc_status.esc[channel].esc_rpm = (erpm * 100) / (_param_mot_pole_count.get() / 2);
+			esc_status.esc[channel].actuator_function = _telemetry->actuator_functions[channel];
```

Comment:
There's still an index mismatch here, it should be:
```cpp
int telemetry_index = 0;
for (channel = 0; channel < _num_outputs; channel++) {
  if (_mixing_output.isFunctionSet(i)) {
    esc_status.esc[telemetry_index].actuator_function = _telemetry->actuator_functions[telemetry_index];
    ...
    ++telemetry_index;
  }
}
```

Reply:
Fixed the numbering logic, now we use sequential numbering for each ESC that provides telemetry data.